Redis client side cluster manager

vipin211 opened the pull request:
- redis queries are executed in the process of the caller
- slots to eredis Pid mapping is kept in ets

vipin211 committed:
Redis client side cluster manager
- redis queries are executed in the process of the caller
- cluster slots and mapping of slots to various nodes in the redis
cluster is kept in a gen_server process

vipin211 commented:
Can I ask you guys to give me high level structural comments on this MR. There could be bugs in the code at this point. I want to incorporate your comments early on and will ask detailed review shortly thereafter after testing and bug fixes.

balusu-murali commented on src/eredis_cluster_manager.erl:
```diff
@@ -0,0 +1,415 @@
+-export([handle_cast/2]).
+-export([handle_info/2]).
+-export([terminate/2]).
+-export([code_change/3]).
```
combine them in a single export? since these are anyways standards callbacks.

NikolaBorisov replied:
Use one export definition with list like our other modules.

balusu-murali commented on src/eredis_cluster_manager.erl:
```diff
@@ -0,0 +1,415 @@
+
+%% Looks up eredis Pid for Slot.
+-spec get_eredis_pid(Cluster::any(), Slot::integer()) -> 
+    {Pid::pid() | undefined, Version::integer()}.
```
why is cluster type any() here? what is it supposed to be? name of the cluster: atom/string/ or something else?

NikolaBorisov commented on src/eredis_cluster_manager.erl:
```diff
@@ -0,0 +1,415 @@
+% API
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+-spec start_link([]) -> {ok, pid()}.
+start_link({ModName, Args}) ->
+    gen_server:start_link({local, ModName}, ?MODULE, [Args], []).
```
fix the spec,
rename ModName to RedisCluster, 
specify Args is list of initial contact nodes.

NikolaBorisov commented on src/eredis_cluster_manager.erl:
```diff
@@ -0,0 +1,415 @@
+%% Looks up eredis Pid for Slot.
+-spec get_eredis_pid(Cluster::any(), Slot::integer()) -> 
+    {Pid::pid() | undefined, Version::integer()}.
+get_eredis_pid(Cluster, Slot) ->
+    {ok, Result} = gen_server:call(Cluster, {get_pid_by_slot, Slot}),
+    Result.
```
usually the name of the function is the same as the message send to the gen_server. Also the result is returned as is. aka if you don't want the ok don't return it from the gen_server

NikolaBorisov commented on src/eredis_cluster_manager.erl:
```diff
@@ -0,0 +1,415 @@
+    {Pid::pid() | undefined, Version::integer()}.
+get_pid_by_slot(State, Slot) ->
+    Index = element(Slot + 1, State#state.slots),
+    Cluster = element(Index, State#state.slots_maps),
```
rename Cluster to SlotMap

NikolaBorisov commented on src/eredis_cluster_manager.erl:
```diff
@@ -0,0 +1,415 @@
+-spec get_cluster_slots([#node{}]) -> [[bitstring() | [bitstring()]]].
+get_cluster_slots([]) ->
+    throw({error,cannot_connect_to_cluster});
+get_cluster_slots([Node|T]) ->
```
when you are trying to get the cluster metadata I would recommend that we do it on a new connection just for this purpose. Trying to reuse the existing cluster connections might be an issue.

vipin211 replied:
I would like to reuse. I think reuse simplifies the code.

NikolaBorisov replied:
Ok we can try this.

balusu-murali commented on src/eredis_cluster_manager.erl:
```diff
@@ -0,0 +1,415 @@
+  [ Index || {_,Index} <- SortedSlotsCache].
+
+
+-spec connect_all_slots([#slots_map{}]) -> [integer()].
+connect_all_slots(SlotsMapList) ->
+    [SlotsMap#slots_map{node=connect_node(SlotsMap#slots_map.node)} || SlotsMap <- SlotsMapList].
+
+
+-spec connect_node(#node{}) -> #node{} | undefined.
+connect_node(Node) ->
+    case lookup_eredis_pid(Node#node.address, Node#node.port) of
+        {ok, _} -> Node;
+        _ -> undefined
+    end.
```
why are we returning `undefined` here if we are unable to connect? shouldn't we log an error?

vipin211 replied:
handled.

NikolaBorisov commented on src/eredis_cluster_manager.erl:
```diff
@@ -0,0 +1,415 @@
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+init_eredis_connections() ->
+    ets:new(?MODULE, [protected, set, named_table, {read_concurrency, true}]).
```
this means you are sharing this table between cluster. I don't think we want this. We should have ets table per cluster.

NikolaBorisov commented on src/eredis_cluster_manager.erl:
```diff
@@ -0,0 +1,415 @@
+%%    end. 
+
+
+safe_eredis_start_link(Ip, Port) ->
+    process_flag(trap_exit, true),
+    Payload = eredis:start_link(Ip, Port),
+    process_flag(trap_exit, false),
+    Payload.
```
Can you explain what this does?

balusu-murali replied:
Here's my understanding:
Executing process_flag in process1 basically traps exits signals from other process that are linked to it. This results in our process1 not crashing when the linked processes exit: since the exit signal is transformed into another message. If we do not trap exit signals: then when the linked process crashes: our process will crash as well.

vipin211 replied:
@balusu-murali explanation is accurate.

I don't understand it fully yet. I used the current code in eredis_cluster as a guideline to write this code. I will spend a bit more time to understand and explain in a day or so.

balusu-murali replied:
They basically do not want this cluster_manager gen_server process to not die if there was any issue with the eredis client process. They probably aim to discover that the eredis client process is not working properly later and attempt to reconnect it.

I think these exit messages need to be handled somewhere?

NikolaBorisov replied:
What we want is our gen_server never to die if one of the eredis connections crashes. We need to trap signals from all the eredis process, or alternatively we should somehow create them as monitor where eredis clients die if we die, but if they die we don't die.

balusu-murali replied:
yes, exactly! that is what I was suggesting. So we need to trap these exit messages from the process that die. I think there probably is some code to do this already in eredis_cluster.erl ?

vipin211 replied:
Added some code to handle these errors. Not perfect. Let me know what you think.

NikolaBorisov commented on src/eredis_cluster_manager.erl:
```diff
@@ -0,0 +1,415 @@
+lookup_eredis_pid(Ip, Port) ->
+    case ets:lookup(?MODULE, [Ip, Port]) of
+        [] ->
+           {ok, Pid} = safe_eredis_start_link(Ip, Port),
+           ets:insert(?MODULE, {[Ip, Port], Pid}),
+          {ok, Pid};
```
indentation

NikolaBorisov commented on src/eredis_cluster_manager.erl:
```diff
@@ -0,0 +1,415 @@
+
+
+init([Args]) ->
+    init_eredis_connections(),
```
should be renamed to create_ets_connection_table

vipin211 replied:
I think done.

balusu-murali commented on src/eredis_cluster_manager.erl:
```diff
@@ -0,0 +1,415 @@
+  SlotsCache = [[{Index, SlotsMap#slots_map.index}
+        || Index <- lists:seq(SlotsMap#slots_map.start_slot, SlotsMap#slots_map.end_slot)]
+        || SlotsMap <- SlotsMaps],
+  SlotsCacheF = lists:flatten(SlotsCache),
```
nit: rename to FlatSlotsCache maybe?

balusu-murali commented on src/eredis_cluster_manager.erl:
```diff
@@ -0,0 +1,415 @@
+
+
+%% Remaps various redis slots.
+-spec remap_cluster(Cluster::any(), Version::integer()) -> {Version::integer()}.
+remap_cluster(Cluster, Version) ->
```
I don't understand this API.. why should the caller mention a version here?
Caller should only ask to remap the cluster if it encounters an error.

vipin211 replied:
Using version allows the code to ignore stale requests. I intend to keep this API for internal use. Only redis queries should ideally be used by the caller. Please let me know your comments.

balusu-murali replied:
yea, I understand. I did not fully see this back then. Thank you vipin!

balusu-murali commented on src/eredis_cluster_manager.erl:
```diff
@@ -0,0 +1,415 @@
+
+-spec throttle_retries(integer()) -> ok.
+throttle_retries(0) -> ok;
+throttle_retries(_) -> timer:sleep(?REDIS_RETRY_DELAY).
```
where is this defined?

vipin211 replied:
in the .hrl file

NikolaBorisov commented:
Generally it is in the right direction. I was imagining we will do a brand new repository, since we are modifying the implementation in very significan way, and the interface changes. I was thinking we will call it ecredis (Erlang Cluster Redis). I think we should also open source it. If we keep the same name as eredis_cluster it will be hard for people to understand it is different and not just fork. We can do this work later I guess. It is nice if the name is short when we want to make a query, something like:

    credis:q(?C, ["SET", ...

vipin211 commented:
Thanks for all the comments. I will incorporate and send modified code in a day or so.

vipin211 committed:
70K ops/sec

vipin211 commented:
Can I ask you to review again? I am looking for API and design level comments. Using the attached perf test, this code delivers ~75K ops/sec, as compared to eredis_cluster(newest version): 46K ops/sec, eredis_cluster_client: 7.5K ops/sec. All three used 3 nodes redis cluster on my laptop. Last two had 20 worker processes. ecredis consumed ~3.6 clients side cores, eredis_cluster consumed ~5.6 client side cores, eredis_cluster_client consumed ~1 client side core.

Once I have addressed all your comments here, I will create a new repo and will continue the review there.

Just FYI: I don't intend to merge this PR in HalloAppInc/eredis_cluster. Please let me know what you think.

NikolaBorisov commented:
> Can I ask you to review again? I am looking for API and design level comments. Using the attached perf test, this code delivers ~75K ops/sec, as compared to eredis_cluster(newest version): 46K ops/sec, eredis_cluster_client: 7.5K ops/sec. All three used 3 nodes redis cluster on my laptop. Last two had 20 worker processes. ecredis consumed ~3.6 clients side cores, eredis_cluster consumed ~5.6 client side cores, eredis_cluster_client consumed ~1 client side core.
> 
> Once I have addressed all your comments here, I will create a new repo and will continue the review there.
> 
> Just FYI: I don't intend to merge this PR in HalloAppInc/eredis_cluster. Please let me know what you think.

@vipin211 The plan looks good. Can I ask you to do the test again on AWS where the client is zone d and the cluster is created by amazon. I'm most interested to see the performance number there because of the 1ms latency to the redis (this is not the same on your laptop). Please compare the 3 cluster clients. Also it will be very useful to see the numbers for 1,2,5,10,20,50,100,200 worker process(generating redis requests). This way we can see it scales higher with more workers while the other 2 clients do not.

NikolaBorisov commented on src/ecredis.erl:
```diff
@@ -0,0 +1,420 @@
+         remap_cluster/2,
+         get_eredis_pid/2,
+         qp/2,
+         q/2]).
```
```suggestion
-export([
    start_link/1,
    remap_cluster/2,
    get_eredis_pid/2,
    qp/2,
    q/2
]).
```

NikolaBorisov commented on src/ecredis.erl:
```diff
@@ -0,0 +1,420 @@
+         q/2]).
+
+%% Callbacks for gen_server.
+-export([init/1,
```
Do this one like the one above, or one line. I sometimes do the gen_server on one line because we are not going to change it. Gen-server is not going to get new callbacks added likely. The other export directive for the module API you are likely to want to add more functions to it, so I like it one per line.

balusu-murali replied:
+1

NikolaBorisov commented on src/ecredis.erl:
```diff
@@ -0,0 +1,420 @@
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+-spec start_link([]) -> {ok, pid()}.
+start_link({ClusterName, Args}) ->
```
```suggestion
-spec start_link(ClusterName :: atom(), InitNodes :: [#node{}]) -> {ok, pid()}.
start_link(ClusterName, InitNodes) ->
```

NikolaBorisov commented on src/ecredis.erl:
```diff
@@ -0,0 +1,420 @@
+
+-spec connect_all_slots(ClusterName::atom(), [#slots_map{}]) -> [integer()].
+connect_all_slots(ClusterName, SlotsMapList) ->
+    [SlotsMap#slots_map{node=connect_node(ClusterName, SlotsMap#slots_map.node)} ||
```
I find something very confusing:
In `parse_cluster_slots` #slot_map.node is of type #node{}, but here you are overriding it and sometimes setting it to undefined... 

Changing the type of what

vipin211 replied:
fixed

NikolaBorisov commented on src/ecredis.erl:
```diff
@@ -0,0 +1,420 @@
+    lists:reverse(Acc).
+
+
+-spec connect_all_slots(ClusterName::atom(), [#slots_map{}]) -> [integer()].
```
add spaces on both sides of `::` everywhere in the code.

NikolaBorisov commented on src/ecredis.erl:
```diff
@@ -0,0 +1,420 @@
+    case Res of
+        {ok, _} -> Node;
+        _ -> undefined
+    end.
```
```suggestion
-spec connect_node(ClusterName::atom(), #node{}) -> ok.
connect_node(ClusterName, Node) ->
    %% Will connect if no connection already present.
    lookup_eredis_pid(ClusterName, Node#node.address, Node#node.port),
    ok.
```

NikolaBorisov commented on src/ecredis.erl:
```diff
@@ -0,0 +1,420 @@
+        true ->
+            undefined
+    end,
+    ets:insert(ets_table_name(State#state.cluster_name), {Slot, {NewPid, State#state.version}}),
```
I think it is confusing to use one ets table for both {Host, Post} -> Pid map and also fro the Slot -> {Pid, version}.
I would have put this into 2 different tables. Also maybe we only need the Slot -> {Pid, version} to be a ets table. Since the {Host, Post -> Pid map is basically only accessed from the gen_server process... I'm fine with 2 ets tables...

balusu-murali replied:
maybe rename the tables to be: `cluster_name_node_pids` and `cluster_name_slot_pids`!

NikolaBorisov commented on src/ecredis.erl:
```diff
@@ -0,0 +1,420 @@
+
+    %% Reset tuples not needed in the state.
+    NewState2 = NewState#state{
+      slots_cache = {},
+      slots_maps = {}
```
It is kind of strange to keep delete this slots_cache and slots_map here. Yes we don't need the slots_cache to be stored in the state, Why don't we just combine the logic in create_slots_cache with cache_eredis_pids and basically process the SlotsMap into the ets tables.

NikolaBorisov commented on src/ecredis.erl:
```diff
@@ -0,0 +1,420 @@
+    case Res of
+        {ok, Pid} ->
+          case eredis:q(Pid, ["CLUSTER", "SLOTS"]) of
+            {error,<<"ERR unknown command 'CLUSTER'">>} ->
```
add space after ,

NikolaBorisov commented on src/ecredis.erl:
```diff
@@ -0,0 +1,420 @@
+    query(ClusterName, Command, Key).
+
+
+query(_Cluster, _, undefined) ->
+    {error, invalid_cluster_key};
```
This way we get more info about what the bad command is.
```suggestion
query(_Cluster, Command, undefined) ->
    {error, invalid_cluster_key, Command};
```

NikolaBorisov commented on src/ecredis.erl:
```diff
@@ -0,0 +1,420 @@
+
+
+query(_Cluster, undefined, _, _, _, ?REDIS_CLUSTER_REQUEST_TTL) ->
+    {error, no_connection};
```
add to the error the slot that is not connected.

NikolaBorisov commented on src/ecredis.erl:
```diff
@@ -0,0 +1,420 @@
+            undefined;
+        "eval" ->
+            get_key_from_rest(Rest);
+        "keys" ->
+            keys;
```
for keys return undefined. we should not allow the keys command.

vipin211 commented:
> > Can I ask you to review again? I am looking for API and design level comments. Using the attached perf test, this code delivers ~75K ops/sec, as compared to eredis_cluster(newest version): 46K ops/sec, eredis_cluster_client: 7.5K ops/sec. All three used 3 nodes redis cluster on my laptop. Last two had 20 worker processes. ecredis consumed ~3.6 clients side cores, eredis_cluster consumed ~5.6 client side cores, eredis_cluster_client consumed ~1 client side core.
> > Once I have addressed all your comments here, I will create a new repo and will continue the review there.
> > Just FYI: I don't intend to merge this PR in HalloAppInc/eredis_cluster. Please let me know what you think.
> 
> @vipin211 The plan looks good. Can I ask you to do the test again on AWS where the client is zone d and the cluster is created by amazon. I'm most interested to see the performance number there because of the 1ms latency to the redis (this is not the same on your laptop). Please compare the 3 cluster clients. Also it will be very useful to see the numbers for 1,2,5,10,20,50,100,200 worker process(generating redis requests). This way we can see it scales higher with more workers while the other 2 clients do not.

https://docs.google.com/spreadsheets/d/17b3Q2mHBHHhH8Dw0wxAUuonwTH9hnLu9nWO6iG1JpWI/edit#gid=0

balusu-murali commented on src/ecredis_sup.erl:
```diff
@@ -0,0 +1,25 @@
+-spec init([])
+	-> {ok, {{supervisor:strategy(), 1, 5}, [supervisor:child_spec()]}}.
+init([]) ->
+    Procs = [{ecredis_a, {ecredis, start_link, [{ecredis_a, [{"127.0.0.1",30001}]}]}, permanent, 5000, worker, [dynamic]},
+             {ecredis_b, {ecredis, start_link, [{ecredis_b, [{"127.0.0.1",30005}]}]}, permanent, 5000, worker, [dynamic]}],
+    {ok, {{one_for_one, 1, 5}, Procs}}.
+
```
split into different lines and indent them properly? makes it easier to read I guess.

balusu-murali commented on src/ecredis.erl:
```diff
@@ -0,0 +1,420 @@
+    ok.
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
```
nit: new line

balusu-murali commented on src/ecredis.erl:
```diff
@@ -0,0 +1,420 @@
+get_key_slot(K) ->
+    %% cast any type of key to a list.
+    Key = lists:concat([K]),
+    KeyToBeHashed = case string:chr(Key,${) of
+                        0 ->
+                            Key;
+                        Start ->
+                            case string:chr(string:substr(Key,Start+1),$}) of
+                                0 ->
+                                    Key;
+                                Length ->
+                                    if
+                                        Length =:= 1 ->
+                                            Key;
+                                        true ->
+                                            string:substr(Key,Start+1,Length-1)
+                                    end
+                            end
+                    end,
```
looks like string:chr is obsolete, please use string:find!

vipin211 replied:
discussed it offline. string:chr is fast. I was considering using string:tokens, which is much slower.
If you have a better suggestion, let me know.

balusu-murali commented on src/ecredis.erl:
```diff
@@ -0,0 +1,420 @@
+        cluster_name = ClusterName
+    },
+    create_ets_table(ClusterName),
+    {ok, connect_(State, InitNodes)}.
```
rename to connect_to_nodes?

balusu-murali commented on src/ecredis.erl:
```diff
@@ -0,0 +1,420 @@
+  end.
+
+
+-spec get_cluster_slots_from_single_node(#node{}) -> [[bitstring() | [bitstring()]]].
```
fix spec to binary(): in all places in the file unless necessary..

vipin211 replied:
I don't understand. Can you give example?

balusu-murali replied:
```suggestion
-spec get_cluster_slots_from_single_node(#node{}) -> [[binary() | [binary()]]].
```

balusu-murali commented on src/ecredis.erl:
```diff
@@ -0,0 +1,420 @@
+    },
+    [cache_eredis_pids(NewState, Slot) || Slot <- lists:seq(0, ?REDIS_CLUSTER_HASH_SLOTS - 1)],
+
+    %% Reset tuples not needed in the state.
```
if they are not needed in the state: then better to avoid storing them in the state at all.

balusu-murali commented on src/ecredis.erl:
```diff
@@ -0,0 +1,420 @@
+%% TODO(vipin): Need to reset connection on Redis node removal.
+-spec reload_slots_map(State::#state{}) -> State::#state{}.
+reload_slots_map(State) ->
+    ClusterSlots = get_cluster_slots(State#state.cluster_name, State#state.init_nodes),
+    SlotsMaps = parse_cluster_slots(ClusterSlots),
+    ConnectedSlotsMaps = connect_all_slots(State#state.cluster_name, SlotsMaps),
+    SlotsCache = create_slots_cache(ConnectedSlotsMaps),
+
+    NewState = State#state{
+        slots_cache = list_to_tuple(SlotsCache),
+        slots_maps = list_to_tuple(ConnectedSlotsMaps),
+        version = State#state.version + 1
+    },
+    [cache_eredis_pids(NewState, Slot) || Slot <- lists:seq(0, ?REDIS_CLUSTER_HASH_SLOTS - 1)],
+
```
I agree with nikola. this logic can be simplified here.
once we have the cluster slots and have connected to these nodes. we can directly call the logic to cache it: and send the slots map to the function and insert it into the ets table.

balusu-murali commented on src/ecredis.erl:
```diff
@@ -0,0 +1,420 @@
+
+%% Looks up existing redis client pid using Ip, Port and return Pid of the client
+%% connection. If no such connection is present, this method establishes the connection.
+-spec lookup_eredis_pid(ClusterName::atom(), Ip::string(), Port::integer()) -> {ok, Pid::pid()}.
+lookup_eredis_pid(ClusterName, Ip, Port) ->
```
nit: I would have passed both Ip and port as a tuple: since functions that are named with lookup usually expected a single key!

balusu-murali commented on src/ecredis.erl:
```diff
@@ -0,0 +1,420 @@
+    end. 
+
+
+%% Remaps various redis slots.
+-spec remap_cluster(ClusterName::atom(), Version::integer()) -> {Version::integer()}.
+remap_cluster(ClusterName, Version) ->
```
callers dont need this api right? because this is already being handled in this query calls?
it's definitely good to have when building out a separate package. 
I was wondering, specifically for our case: do you think we will use this?

vipin211 replied:
We could force the cluster to remap instead of waiting for it to happen automatically.

balusu-murali reviewed (commented):
Nice work vipin!
This change was hard to review with lots of new code. I would have first created a new file with the old cluster client (because core functions like cluster_slots and others are still the same) and then I would have made custom changes: it could make things a lot easier to review.

vipin211 commented:
> Nice work vipin!
> This change was hard to review with lots of new code. I would have first created a new file with the old cluster client (because core functions like cluster_slots and others are still the same) and then I would have made custom changes: it could make things a lot easier to review.

I agree. What you see in the PR is WIP to get correctness and performance. I will address comments and refactor shortly.

NikolaBorisov commented:
Performance looks great it is 70x faster basically.

vipin211 committed:
Address review comments

vipin211 committed:
Handle eredis:start_link failure

vipin211 committed:
Review comments

vipin211 reviewed (commented):
Addressed all the comments. Please take a look again.

balusu-murali commented on src/ecredis_sup.erl:
```diff
@@ -0,0 +1,31 @@
+start_link() ->
+    supervisor:start_link({local, ?MODULE}, ?MODULE, []).
+
+-spec init([])
+	-> {ok, {{supervisor:strategy(), 1, 5}, [supervisor:child_spec()]}}.
```
nit: spec can be on the same line I guess.

balusu-murali commented on src/ecredis.erl:
```diff
@@ -0,0 +1,431 @@
+-define(SLOT_PIDS, slot_pids).
+
+%% API.
+-export([start_link/2,
+         remap_cluster/2,
+         get_eredis_pid/2,
+         qp/2,
+         q/2
+]).
```
```suggestion
-export([
    start_link/2,
    remap_cluster/2,
    get_eredis_pid/2,
    qp/2,
    q/2
]).
```

balusu-murali commented on src/ecredis.erl:
```diff
@@ -0,0 +1,431 @@
+-define(SLOT_PIDS, slot_pids).
+
+%% API.
+-export([start_link/2,
```
how about another api: `get_eredis_pid_with_key`: this can take in the key and the cluster name.. then hash the key, get the slot and get the pid based on the slot? let me know what you think?

vipin211 replied:
I can add it in future.

balusu-murali commented on src/ecredis.erl:
```diff
@@ -0,0 +1,431 @@
+
+-spec remap_cluster_internal(State :: #state{}, Version :: integer()) -> State :: #state{}.
+remap_cluster_internal(State, Version) ->
+    if State#state.version == Version ->
+        reload_slots_map(State);
+    true ->
+        State
+    end.
```
nit:
```suggestion
    if
        State#state.version == Version ->
            reload_slots_map(State);
        true ->
            State
    end.
```

balusu-murali commented on src/ecredis.erl:
```diff
@@ -0,0 +1,431 @@
+reload_slots_map(State) ->
+    ClusterSlots = get_cluster_slots(State#state.cluster_name, State#state.init_nodes),
+    SlotsMaps = parse_cluster_slots(ClusterSlots),
+    connect_all_slots(State#state.cluster_name, SlotsMaps),
+    NewState = State#state{
+        version = State#state.version + 1
+    },
```
Small suggestion: I would have given the state as an input to the connect_all_slots function. That should connect everything, upgrade the version and return the state back. something like this:
```
NewState = connect_all_slots(SlotsMaps, State)
```
I saw this code a lot in ejabberd modules and liked it.. so wanted to mention. You need not change anything if you dont like it.

balusu-murali commented:
@vipin211 : nice work! the performance results are great!
Looks good overall. I was halfway through this.. will review your new pr.

vipin211 reviewed (commented):
@balusu-murali I addressed the comments and sent the update to the new repo.

NikolaBorisov commented:
@vipin211 we are abandoning this PR, and we are doing new repository right?

vipin211 commented:
> @vipin211 we are abandoning this PR, and we are doing new repository right?

Yes. But if you think I have not addressed your comment in this PR, don't hesitate to add your comments here or on the new PR and I will address them in the new PR: https://github.com/HalloAppInc/ecredis/pull/1. Let me know if you are not ok with it.

mdonohue commented:
I think there's a way to mark a PR as closed or obsolete, which is helpful for future viewers.

vipin211 commented:
Please use this PR for future comments: https://github.com/HalloAppInc/ecredis/pull/1. Closing this PR.
